crates/ty_python_semantic/resources/mdtest/deprecated.md

```diff
@@ -0,0 +1,256 @@
+# Tests for the `@deprecated` decorator
+
+## Introduction
+
+The decorator `@deprecated("some message")` can be applied to functions, methods, overloads, and
+classes. Uses of these items should subsequently produce a warning.
+
+```py
+from typing_extensions import deprecated
+
+@deprecated("use OtherClass")
+def myfunc(): ...
+
+myfunc()  # error: [deprecated] "use OtherClass"
+```
+
+```py
+from typing_extensions import deprecated
+
+@deprecated("use BetterClass")
+class MyClass: ...
+
+MyClass()  # error: [deprecated] "use BetterClass"
+```
+
+```py
+from typing_extensions import deprecated
+
+class MyClass:
+    @deprecated("use something else")
+    def afunc(): ...
+    @deprecated("don't use this!")
+    def amethod(self): ...
+
+MyClass.afunc()  # error: [deprecated] "use something else"
+MyClass().amethod()  # TODO error: [deprecated] "don't use this!"
+```
+
+## Syntax
+
+The decorator takes one mandatory positional string argument (the message) and two optional
+arguments that affect runtime behaviour but don't affect static checks.
+
+```py
+from typing_extensions import deprecated
+
+@deprecated
+def invalid_deco(): ...
+
+invalid_deco()  # error: [missing-argument]
+```
+
+```py
+from typing_extensions import deprecated
+
+x = "message"
+
+@deprecated(x)
+def invalid_deco(): ...
+
+invalid_deco()  # error: [deprecated] "message"
+```
+
+We currently don't verify the other arguments.
+
+```py
+from typing_extensions import deprecated
+
+@deprecated("some message", dsfsdf="whatever")
+def invalid_deco(): ...
+
+invalid_deco()  # error: [deprecated] "some message"
+```
+
+But we should always handle correct ones fine.
+
+```py
+from typing_extensions import deprecated
+
+@deprecated("some message", category=DeprecationWarning, stacklevel=1)
+def valid_deco(): ...
+
+valid_deco()  # error: [deprecated] "some message"
+```
+
+## Different Versions
+
+There are 3 different sources of `@deprecated`: `warnings`, `typing`, and `typing_extensions`. The
+version in `warnings` is canoninical, the other versions are compatibility shims.
+
+`main.py`:
+
+```py
+import warnings
+import typing
+import typing_extensions
+
+# TODO(why is mdtest not finding this?) @warnings.deprecated("nope")
```

Comment:
It's only been added very recently in 3.13: https://github.com/python/typeshed/blob/main/stdlib/warnings.pyi#L120-L121

Mdtests run on 3.9 by default, but you can change this by adding a `toml`-block in this section to switch the version to 3.13: https://github.com/astral-sh/ruff/blob/main/crates/ty_test/README.md#configuration